subprojects/core/src/main/java/org/gradle/initialization/DefaultProjectDescriptor.java

```diff
@@ -41,6 +41,7 @@
     public static final String BUILD_SCRIPT_BASENAME = "build";
 
     private String name;
+    private boolean nameExplicitlySet; // project name explicitly specified in the build script (as opposed to derived from the containing folder)
```

Comment:
We could also set this boolean from the constructor if the name is different than the path passed in. It would allow a value restored from the configuration cache to be correct in more cases.

Reply:
I'm not sure it's a good idea. On the one hand, the whole logic doesn't trigger with CC on, on the other hand, what about the situation when the root project name is explicitly set and also matches the path? We would be just exchanging false positives with false negatives. 

With the current logic I know that if the flag is `true` then the name has been explicitly set for sure. So I err on the side of warning, but with the more complicated logic I might not warn when I should.

Reply:
Ok makes sense